numpy/random/generator.pyx

```diff
@@ -3857,8 +3866,12 @@ cdef class Generator:
                     buf[...] = x[j]
                     x[j] = x[i]
                     x[i] = buf
+            x = np.swapaxes(x, 0, axis)
         else:
             # Untyped path.
+            axis = normalize_axis_index(axis, np.ndim(x))
+            if axis != 0:
+                raise np.AxisError(axis, x.ndim)
```

Comment:
This doesn't make any sense to me - the message it prints out will just be wrong if a.ndim isn't 1.

Reply:
@eric-wieser It seems for untyped path, the function only shuffle on `0` axis?

Reply:
@eric-wieser But you are right, the error message is not correct. Should we raise another error?

Reply:
I think on this path, you should use:
```
if axis != 0:
    x = np.swapaxes(x, ...)
```

That way, if `swapaxes` is supported on a duck-array, it will just work. If it isn't supported, then the duckarray will be able to report itself that its not supported.

By adding the `!=0` check, you ensure that duckarrays don't need to implement swapaxes to support axis=0.

Reply:
@eric-wieser I see. If `swapaxes` works, we need to add another `swapaxes` after the block`with self.lock ...` ?

Reply:
Correct